### PR TITLE
vstart.sh: add options to set number of alien threads, and number of cpu cores for alien threads

### PR DESCRIPTION
**Pull request by perezjosibm**
### [vstart.sh] 
Add two new command line options to

1. set number of alien threads,` --crimson-alien-num-threads `, other than the default value(6),
2. set number of cpu cores for alien threads,  `--crimson-alien-num-cores` -- the default is to use all the other cores in the system not used by crimson/seastar.

This is an example of the added functionality, and the verification via inspection at the debug output, as well as via ps below
```
# MDS=0 MON=1 OSD=1 MGR=1 ../src/vstart.sh --new -x --localhost --without-dashboard -o timeout=10000\
 -o session_timeout=10000 --bluestore --crimson --redirect-output --bluestore-devs /dev/sdc --crimson-smp 4\
 --crimson-alien-num-threads 2 --crimson-alien-num-cores 1

DEBUG 2024-04-15 15:34:41,816 [shard 0:main] monc - set_mon_vals crimson_alien_op_num_threads = 2
DEBUG 2024-04-15 15:34:41,816 [shard 0:main] monc - set_mon_vals crimson_alien_thread_cpu_cores = 4-4 <- correct!
DEBUG 2024-04-15 15:34:41,816 [shard 0:main] monc - set_mon_vals crimson_seastar_cpu_cores = 0-3

# ps -p $(pgrep crimson-osd) -L -o pid,tid,comm,psr,pcpu
    PID     TID COMMAND         PSR %CPU
 144843  144843 crimson-osd       0  0.4
 144843  144851 reactor-1         1  0.3
 144843  144852 reactor-2         2  0.3
 144843  144853 reactor-3         3  0.3
 144843  144854 syscall-0         0  0.0
 144843  144855 syscall-2         2  0.0
 144843  144856 syscall-3         3  0.0
 144843  144857 syscall-1         1  0.0
 144843  144858 log               0  0.0
 144843  144859 alien-store-tp    4  0.0
 144843  144860 alien-store-tp    4  0.0
 144843  144861 crimson-osd       0  0.0
 144843  144862 reactor-3         3  0.0
 144843  144863 reactor-1         1  0.0
 144843  144864 reactor-2         2  0.0
 144843  144865 bstore_aio        4  0.0
 144843  144867 rocksdb:low       4  0.0
 144843  144868 rocksdb:low       4  0.0
 144843  144869 rocksdb:low       4  0.0
 144843  144870 rocksdb:high      4  0.0
 144843  145051 bstore_aio        4  0.0
 144843  145232 alien-store-tp    4  0.0
 144843  145233 cfin              4  0.0
 144843  145234 bstore_kv_sync    4  0.0
 144843  145235 bstore_kv_final   4  0.0
 144843  145236 bstore_mempool    4  0.0
```
Note: an extra alien thread has always been created, no change to that internal aspect has been changed.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/vstart.sh [hunk 1]
```diff
@@ -1697,8 +1715,15 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
         extra_seastar_args=" --trace"
     fi
     if [ "$(expr $(nproc) - 1)" -gt "$(($CEPH_NUM_OSD * crimson_smp))" ]; then
-      echo "crimson_alien_thread_cpu_cores:" $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
-      $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
+        if [ $crimson_alien_num_cores -gt 0 ]; then
+            alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
+            alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
+            echo "crimson_alien_thread_cpu_cores: $alien_bottom_cpu-$alien_top_cpu"
+            $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores "$alien_bottom_cpu-$alien_top_cpu"
```
Comment (cyx1231st) [inline src/vstart.sh:1722]: If there are 2 CPU sockets, each with 16 physical cores and are hyper-threaded, the range might be:
CPU0: 0-15, 32-47
CPU1: 16-31,48-63

This means the range of logical cpu cores from the physical ones may not be consecutive. If want to pin the seastar threads to 8 physical cores (CPU0 0-7), and pin alien threads to the rest of 8 physical cores on CPU0, the proper way might need to set them to 8-15,40-47.

What do you think?

**Comment by cyx1231st** [comment]
> This is an example of the added functionality, and the verification via inspection at the debug output, as well as via ps below

I have also observed the same taskset results, that all the bluestore threads `rocksdb:low` `rocksdb:high` `bstore_aio` etc seem to be pinned correctly (to the same range of the configuration `crimson_alien_thread_cpu_cores`).

However, looking at the source code, I can only see this configuration is pinning the `alien-store-tp` threads from the `ThreadPool`:
https://github.com/ceph/ceph/blob/d50b94f93f6f94a8bccca88a5bf796efdb18e9d8/src/crimson/os/alienstore/thread_pool.cc#L43-L53

My concern is that the current way is happened to be correct due to some unknown side-effects, which is subtle and can become unexpected anytime.

File: src/vstart.sh (hunk 1 above)
Comment (Matan-B) [reply src/vstart.sh:1722]: I think that if we need a precise cpu allocation we can manually set `crimson_alien_thread_cpu_cores` (and override the "primitive" ones already set here by vstart).
The new `crimson_alien_num_cores` vstart flag is introduced only to enable limiting of cores used IIUC.

File: src/vstart.sh [hunk 2]
```diff
@@ -1142,6 +1156,10 @@ start_osd() {
 	    # set exclusive CPU nodes for each osd
 	    echo "$CEPH_BIN/ceph -c $conf_fn config set osd.$osd crimson_seastar_cpu_cores $bottom_cpu-$top_cpu"
 	    $CEPH_BIN/ceph -c $conf_fn config set "osd.$osd" crimson_seastar_cpu_cores "$bottom_cpu-$top_cpu"
+        if [ $crimson_alien_num_threads -gt 0 ]; then
+	        echo "$CEPH_BIN/ceph -c $conf_fn config set osd.$osd crimson_alien_op_num_threads $crimson_alien_num_threads"
+	        $CEPH_BIN/ceph -c $conf_fn config set "osd.$osd" crimson_alien_op_num_threads "$crimson_alien_num_threads"
+        fi
```
Comment (Matan-B) [inline src/vstart.sh:1162]: Since the value set here is the same for all Crimson OSDs it shouldn't be set for each OSD.
`crimson_alien_op_num_threads` is global to all of the OSDs:
```
$CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_op_num_threads "$crimson_alien_num_threads"
```

File: src/script/cpu-map.sh [hunk 3]
```diff
@@ -0,0 +1,158 @@
+#!/bin/bash
```
Comment (Matan-B) [inline src/script/cpu-map.sh:1]: Can you please move the new script into a separate commit?

File: src/vstart.sh (hunk 1 above)
Comment (cyx1231st) [reply src/vstart.sh:1722]: > The new crimson_alien_num_cores vstart flag is introduced only to enable limiting of cores used IIUC.

I agree, and it might be a step further to improve this limitation based on physical cores rather than the logical ones.

Comment (perezjosibm) [reply src/vstart.sh:1722]: I see, I was wondering about that, whether the alien threads required to be pinned to the "closer" (that is, in the same socket preferably) as the I/O reactor/seastar threads.
What would the general rule be? In principle, pin always on physical cores, in the same socket (if available) to avoid excessive memory bandwidth between sockets. Anything else?

File: src/vstart.sh (hunk 2 above)
Comment (perezjosibm) [reply src/vstart.sh:1162]: Ah, thanks for clarifying, I was not aware of that 👍

File: src/script/cpu-map.sh (hunk 3 above)
Comment (perezjosibm) [reply src/script/cpu-map.sh:1]: Sure, will do

File: src/vstart.sh (hunk 1 above)
Comment (cyx1231st) [reply src/vstart.sh:1722]: > Would be preferable if given a seastar thread pinned to a physical core X, its associated alien(s) thread to be pinned to the hyperthreaded sibling of core X?

My instinct is that alien threads (blockable) should not share the same physical cores with seastar threads (nonblocking). Seastart threads only do nonblock executions, but hyperthreading won't duplicate the main execution resources. I think this is also why seastar threads uses the one-thread-per-core model. Other than that, closer should be better.

Comment (perezjosibm) [reply src/vstart.sh:1722]: Here is a comparison of the effect of modifying the number of alien threads, and number of cores to alien threads (from the example shown in the Description):

```bash
# MDS=0 MON=1 OSD=1 MGR=1 ../src/vstart.sh --new -x --localhost --without-dashboard -o timeout=10000 -o session_timeout=10000 --bluestore --crimson --redirect-output --bluestore-devs /dev/sdc --crimson-smp 4 --crimson-alien-num-threads 2 --crimson-alien-num-cores 1
```

The table on the left shows the reference baseline configuration results; whereas the table on the right hand side, presents the corresponding results on the new configuration with a single core for alien threads. 
Note the values  `(iodepth x num_jobs)`. the Multiple FIO case used 8 processes, each exercising its own RBD image.
Screenshot 2024-05-20 at 12 24 11


Is interesting to note the impact on performance of the new configuration. For example, the following takes a closer look to random read 4k, where the new configuration impacts approximately 25% in throughput: (each row represents the output of a single FIO process) 

Screenshot 2024-05-20 at 12 24 47

Comment (cyx1231st) [reply src/vstart.sh:1723]: > I see, so the current logic should change: currently, we first allocate and pin each I/O reactor to its CPU core, then we allocate the alien threads (they can use any available CPU core left). But now, we need to allocate and pin each sibling alien threads for each I/O reactor (aka seastar) thread, on physical cores only.

In a system of 2 CPU sockets, 8 cores per socket, 2 hyperthreaded logical cores per physical core, the CPU range from OS perspective might be:
* CPU0: 0-7,16-23
* CPU1: 8-15,24-31

As for p2l mapping, e.g. the physical core p0 is mapped to logical [0,16], p1 to l[1,17], ..., p8 to [8,24], etc. (refer to `/proc/cpuinfo`).

If there are 2 OSDs, `--crimson-smp`==6 (each OSD has 6 I/O reactor threads, totally 12), alien threads is set to 4 per OSD (totally 8).

If I understand correctly, the current logic will allocate logical cores [0-5] to OSD0 reactors, [6-11] to OSD1 reactors, and the rest of logical cores [12-31] (probably with a limit from `crimson_alien_num_cores`) to the 8 alien threads.

I think the issue of the current logical-core-based arrangement is the range of alien threads [12-31] means they are allowed to run on the physical cores already running the reactor threads, i.e. [16-27] is shared by p0-11.

Towards physical-core-based arrangement, the only change here might be setting the rest of **physical** cores to the 8 alien threads, which are the logical ones [12-23,28-31] (Alien threads are blockable, so they should be able to benefit from hyperthreading).

As @Matan-B suggested, it is also possible to manually set `crimson_alien_thread_cpu_cores` for the physical-core-based cpu allocation. Also, the impact from the inaccurate policy might be negligible in reality. 

> Other than that, closer should be better.

I think this is actually challenging for the best strategy because alien threads also talk to each other and the storage device, and reactor threads also talk to each other and the network device.

File: src/vstart.sh (hunk 2 above)
Comment (Matan-B) [reply src/vstart.sh:1163]: Can you please update the commit to use the global osd value?

Comment (perezjosibm) [reply src/vstart.sh:1163]: Sure, will do it asap

File: src/vstart.sh (hunk 1 above)
Comment (cyx1231st) [reply src/vstart.sh:1723]: I think the current PR looks good enough, and as @Matan-B suggested, it is possible to configure `-o "crimson_alien_thread_cpu_cores=xxx"` to experiment advanced core allocations (even with multiple OSDs).

(So the below discussion may not be related to this PR)

Sorry I'm not able to follow the above graph. The typical arrangement of (2 CPU sockets, 8 physical cores per socket, 2 hyperthreads per physical core) should look like below (from `lscpu`):
```
...
NUMA:
  NUMA node(s):          2
  NUMA node0 CPU(s):     0-7,16-23
  NUMA node1 CPU(s):     8-15,24-31
...
```
Node0 means socket0, and node1 means socket1.

And I'm not sure why the policy on the left (socket0) can be different from right (socket1) at the same time, while it looks that all the threads come from the same OSD with 8 reactor-threads and 8 alien-threads.

> for each socket, let n be the number of physical cores per socket (assume that each socket has the same n number of hyper-threading cores)

I think it is incorrect to assume that there are n physical cores and n extra hyper-threading cores. From OS perspective, if HT is enabled, all the vCPUs are hyperthreaded: in the above example, vCPU0 and vCPU16 may be symmetric and are both associated to the same physical core in socket0.

Or probably I'm missing something?

Comment (perezjosibm) [reply src/vstart.sh:1723]: Sorry for my abuse of the notation (which caused the confusion), what I meant was the following: I think that the vstart.sh script uses the _processor id_ to identify a cpu to pin the corresponding thread t. The processor id is the second column from the `/proc/cpuinfo `output, I abused the notation to differentiate between the _physical_ and the _hyperthreaded_, from the lscpu output, I called cpus 0 to 7 as physical, and cpus 16-23 hyperthreaded, so _n_ in this case is 8 (because we have the same number of physical and hyperthreading).

```bash
core id  : 0 processor : 0 physical id : 0
:
core id  : 12 processor : 7 physical id : 0
core id  : 0 processor : 8 physical id : 1
:
core id  : 12 processor : 15 physical id : 1
core id  : 0 processor : 16 physical id : 0 <- HT sibling of processor 0
:
core id  : 12 processor : 23 physical id : 0
core id  : 0 processor : 24 physical id : 1 <- HT sibling of processor 8
:
```

In other words, I was suggesting that the script only use the list of _processor id_ [ 0-15], but do not use the ids [16-31]. That was the meaning in the tables above. I was only suggesting two approaches:

-  to allocate an alien thread "close" to its seastar/reactor partner (assuming a 1-1 correspondence, which might not be the optimal case), or 

- to pin first all the seastar/reactors to the available "physical" [0-15], and then to pin the alien threads to the remaining "physical" processor ids -- again, from the list [0-15], if any left.  I guess in the case of 16 OSD, we need to pin the alien threads to HT processor ids because none of the [0-15] have been left -- unless we set that the max number of seastar/reactor threads for the above [0-15] case is 8.

I hope this makes sense 😃

**Comment by perezjosibm** [comment]
Hi @Matan-B and @cyx1231st, I've added a commit for the requested change, but I've not managed to figure out why the ceph API test fail, would you please help me figure out if I missed something? Thanks

**Comment by athanatos** [comment]
@perezjosibm Apologies for the noise -- the api test failure isn't related to this PR.  The test is just annoyingly flaky.  My next comment will cause it to rerun.

**Comment by athanatos** [comment]
jenkins test api

File: src/vstart.sh [hunk 4]
```diff
@@ -1720,7 +1716,12 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
             alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
             alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
             echo "crimson_alien_thread_cpu_cores: $alien_bottom_cpu-$alien_top_cpu"
+            # This is a (logical) processor id range, it could be refined to encompass only physical processor ids
+            # (equivalently, ignore hyperthreading sibling processor ids)
             $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores "$alien_bottom_cpu-$alien_top_cpu"
+            # Setting the global value for the number of alien threads
```
Comment (Matan-B) [inline src/vstart.sh:1722]: I'm not sure that this comment provides any additional information

File: src/vstart.sh [hunk 5]
```diff
@@ -1698,8 +1712,20 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
...
+        if [ $crimson_alien_num_cores -gt 0 ]; then
+            alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
+            alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
+            echo "crimson_alien_thread_cpu_cores: $alien_bottom_cpu-$alien_top_cpu"
+            # This is a (logical) processor id range, it could be refined to encompass only physical processor ids
+            # (equivalently, ignore hyperthreading sibling processor ids)
+            $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores "$alien_bottom_cpu-$alien_top_cpu"
+            # Setting the global value for the number of alien threads
+	        echo "$CEPH_BIN/ceph -c $conf_fn config set osd.$osd crimson_alien_op_num_threads $crimson_alien_num_threads"
+	        $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_op_num_threads "$crimson_alien_num_threads"
```
Comment (Matan-B) [inline src/vstart.sh:1724]: The condition above is: `if [ $crimson_alien_num_cores -gt 0 ]; then`, we should be able to set `crimson_alien_op_num_threads` regardless. Let's add a separate condition for it:
```
if [ $crimson_alien_num_threads -gt 0 ]; then
	echo "$CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_op_num_threads $crimson_alien_num_threads"
	$CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_op_num_threads "$crimson_alien_num_threads"
fi
```

File: src/vstart.sh [hunk 6]
```diff
@@ -1698,8 +1712,20 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
         extra_seastar_args=" --trace"
     fi
     if [ "$(expr $(nproc) - 1)" -gt "$(($CEPH_NUM_OSD * crimson_smp))" ]; then
-      echo "crimson_alien_thread_cpu_cores:" $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
-      $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
+        if [ $crimson_alien_num_cores -gt 0 ]; then
+            alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
+            alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
```
Comment (Matan-B) [inline src/vstart.sh:1717]: `alien_top_cpu` might be out of `nproc` bounds

File: src/vstart.sh [hunk 7]
```diff
@@ -1720,7 +1716,12 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
             alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
             alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
             echo "crimson_alien_thread_cpu_cores: $alien_bottom_cpu-$alien_top_cpu"
+            # This is a (logical) processor id range, it could be refined to encompass only physical processor ids
+            # (equivalently, ignore hyperthreading sibling processor ids)
             $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores "$alien_bottom_cpu-$alien_top_cpu"
+            # Setting the global value for the number of alien threads
+	        echo "$CEPH_BIN/ceph -c $conf_fn config set osd.$osd crimson_alien_op_num_threads $crimson_alien_num_threads"
```
Comment (Matan-B) [inline src/vstart.sh:1723]: ~osd.$osd~ osd

**Review by Matan-B — COMMENTED** [review]
LGTM overall, left a few comments. Please squash the commits.

File: src/vstart.sh (hunk 6 above)
Comment (perezjosibm) [reply src/vstart.sh:1717]: Yes, sorry, I'm fixing adding a condition to set alien_top_cpu to $(nproc) - 1 in such a case, thank you!

**Comment by perezjosibm** [comment]
Testing: latest on SV1-ceph3:
```bash
[root@3564b7abc454 build]# MDS=0 MON=1 OSD=1 MGR=1 ../src/vstart.sh --new -x --localhost --without-dashboard\
 -o timeout=10000 -o session_timeout=10000 --bluestore --redirect-output\
 --bluestore-devs /dev/sdd,/dev/sde,/dev/sdf --crimson  --crimson-smp 4 --crimson-alien\
-num-threads 6 --crimson-alien-num-cores 1

:
crimson_alien_thread_cpu_cores: 4-4
/ceph/build/bin/ceph -c /ceph/build/ceph.conf config set osd crimson_alien_op_num_threads 6
creating /ceph/build/dev/mgr.x/keyring
:
/ceph/build/bin/ceph -c /ceph/build/ceph.conf config set osd.0 crimson_seastar_cpu_cores 0-3
1+0 records in
1+0 records out
:
DEBUG 2024-06-11 17:48:02,008 [shard 0:main] monc - set_mon_vals crimson_alien_op_num_threads = 6
DEBUG 2024-06-11 17:48:02,008 [shard 0:main] monc - set_mon_vals crimson_alien_thread_cpu_cores = 4-4
DEBUG 2024-06-11 17:48:02,008 [shard 0:main] monc - set_mon_vals crimson_seastar_cpu_cores = 0-3
```
visual verification that the number of alien threads is correct:
```bash
[root@3564b7abc454 build]# ps -p $(pgrep crimson-osd) -L -o pid,tid,comm,psr,pcpu
    PID     TID COMMAND         PSR %CPU
   1599    1599 crimson-osd       0  0.7
   1599    1608 reactor-1         1  0.4
   1599    1609 reactor-2         2  0.4
   1599    1610 reactor-3         3  0.4
   1599    1611 syscall-0         0  0.0
   1599    1612 syscall-2         2  0.0
   1599    1613 syscall-3         3  0.0
   1599    1614 syscall-1         1  0.0
   1599    1615 log               0  0.0
   1599    1616 alien-store-tp    4  0.0
   1599    1617 alien-store-tp    4  0.0
   1599    1618 alien-store-tp    4  0.0
   1599    1619 alien-store-tp    4  0.0
   1599    1620 alien-store-tp    4  0.0
   1599    1621 alien-store-tp    4  0.0
   1599    1622 crimson-osd       0  0.0
   1599    1623 reactor-3         3  0.0
   1599    1624 reactor-1         1  0.0
   1599    1625 reactor-2         2  0.0
   1599    1626 bstore_aio        4  0.0
   1599    1628 rocksdb:low       4  0.0
   1599    1629 rocksdb:low       4  0.0
   1599    1630 rocksdb:low       4  0.0
   1599    1631 rocksdb:high      4  0.0
   1599    1812 bstore_aio        4  0.0
   1599    1994 alien-store-tp    4  0.0
   1599    1995 cfin              4  0.0
   1599    1996 bstore_kv_sync    4  0.0
   1599    1997 bstore_kv_final   4  0.0
   1599    1998 bstore_mempool    4  0.0

# ps -p $(pgrep crimson-osd) -L -o pid,tid,comm,psr,pcpu | grep -c alien
7

```
quick test (using RBD since FIO/CBT is not installed on  this container image 😭 ):

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test api

File: src/vstart.sh (hunk 1 above)
Comment (cyx1231st) [reply src/vstart.sh:1725]: Thanks for clarifying!

> * to allocate an alien thread "close" to its seastar/reactor partner (assuming a 1-1 correspondence, which might not be the optimal case), or

It doesn't seem reasonable to me: for 1 OSD with m reactor threads and n alienstore threads, I think a reactor thread is possible talk to all the alienstore threads, and the n alienstore threads may share resources with each other.

> * to pin first all the seastar/reactors to the available "physical" [0-15], and then to pin the alien threads to the remaining "physical" processor ids -- again, from the list [0-15], if any left. I guess in the case of 16 OSD, we need to pin the alien threads to HT processor ids because none of the [0-15] have been left -- unless we set that the max number of seastar/reactor threads for the above [0-15] case is 8.

Yeah, sharing the physical cores may be inevitable in some cases -- in case of 16 physical cores for 16 crimson OSDs, probably the better way is to use a better CPU(s) if the workload is CPU-bounded.

**Review by cyx1231st — APPROVED** [review]
